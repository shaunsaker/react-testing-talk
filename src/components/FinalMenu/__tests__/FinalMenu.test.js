import React from "react";
import renderer from "react-test-renderer";

import { Menu } from "..";

describe("Menu", () => {
  const spies = [];
  const dispatch = jest.fn();
  const handleSelectOption = jest.fn();
  const options = ["Option 1", "Option 2", "Option 3"];
  const headingText = "What is your favourite song?";

  describe("renders", () => {
    it("renders with minimum required props", () => {
      const component = renderer.create(<Menu options={options} />);

      expect(component).toMatchSnapshot();
    });

    it("renders the correct UI based on props", () => {
      const component = renderer.create(
        <Menu headingText={headingText} options={options} />
      );
      const { root } = component;
      const targetComponent = root.findByProps({
        testID: "Menu.headingText"
      });

      expect(targetComponent).toBeDefined();
    });

    it("renders the correct UI based on state", () => {
      const selectedOption = options[1];
      const component = renderer.create(
        <Menu
          headingText={headingText}
          selectedOption={selectedOption}
          options={options}
          dispatch={dispatch}
        />
      );

      expect(component).toMatchSnapshot();
    });
  });

  describe("methods", () => {
    it("should work", () => {
      spies[0] = jest.spyOn(Menu.prototype, "setSelectedOption"); // NB that this is called before creating comoponent
      const component = renderer.create(
        <Menu
          options={options}
          dispatch={dispatch}
          handleSelectOption={handleSelectOption}
        />
      );
      const instance = component.getInstance();
      const selectedOption = options[1];

      instance.onSelectOption(selectedOption);

      expect(spies[0]).toHaveBeenCalledWith(selectedOption);
      expect(handleSelectOption).toHaveBeenCalledWith(selectedOption);
      expect(dispatch).toHaveBeenCalled();
      expect(dispatch).toMatchSnapshot();
    });

    it("should update state correctly", () => {
      const component = renderer.create(<Menu options={options} />);
      const instance = component.getInstance();
      const selectedOption = options[2];

      instance.setSelectedOption(selectedOption);

      expect(instance.state.selectedOption).toEqual(selectedOption);
    });
  });

  describe("lifecycle methods", () => {
    it("should handle componentDidMount correctly", () => {
      spies[0] = jest.spyOn(Menu.prototype, "onSelectOption");
      const selectedOption = options[1];
      renderer.create(
        <Menu selectedOption={selectedOption} options={options} />
      );

      expect(spies[0]).toHaveBeenCalledWith(selectedOption);
    });

    it("should handle componentDidUpdate correctly", () => {
      spies[0] = jest.spyOn(Menu.prototype, "onSelectOption");
      const initialOption = options[1];
      const component = renderer.create(
        <Menu selectedOption={initialOption} options={options} />
      );

      expect(spies[0]).toHaveBeenCalledWith(initialOption);

      spies[0].mockClear();

      const nextOption = options[2];

      component.update(<Menu selectedOption={nextOption} options={options} />);

      expect(spies[0]).toHaveBeenCalledWith(nextOption);
    });
  });

  describe("actions", () => {
    it("should call onSelectOption on menu option button click", () => {
      spies[0] = jest.spyOn(Menu.prototype, "onSelectOption");
      const component = renderer.create(<Menu options={options} />);
      const { root } = component;
      const optionIndex = 2;
      const targetComponent = root.findByProps({
        testID: `Menu.button.${optionIndex}`
      });

      targetComponent.props.onClick();

      expect(spies[0]).toHaveBeenCalledWith(options[optionIndex]);
    });
  });

  afterEach(() => {
    spies.forEach(spy => {
      if (spy) {
        spy.mockClear();
      }
    });
    dispatch.mockClear();
  });
});
