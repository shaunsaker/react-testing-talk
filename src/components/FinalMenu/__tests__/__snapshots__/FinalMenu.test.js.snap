// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu methods should work 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "payload": Object {
          "option": "Option 2",
        },
        "type": "someAction",
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`Menu renders renders the correct UI based on state 1`] = `
<div>
  <h1
    testID="Menu.headingText"
  >
    What is your favourite song?
  </h1>
  <ul>
    <li>
      <button
        onClick={[Function]}
        style={
          Object {
            "color": "black",
          }
        }
        testID="Menu.button.0"
        type="button"
      >
        Option 1
      </button>
    </li>
    <li>
      <button
        onClick={[Function]}
        style={
          Object {
            "color": "red",
          }
        }
        testID="Menu.button.1"
        type="button"
      >
        Option 2
      </button>
    </li>
    <li>
      <button
        onClick={[Function]}
        style={
          Object {
            "color": "black",
          }
        }
        testID="Menu.button.2"
        type="button"
      >
        Option 3
      </button>
    </li>
  </ul>
</div>
`;

exports[`Menu renders renders with minimum required props 1`] = `
<div>
  <ul>
    <li>
      <button
        onClick={[Function]}
        style={
          Object {
            "color": "black",
          }
        }
        testID="Menu.button.0"
        type="button"
      >
        Option 1
      </button>
    </li>
    <li>
      <button
        onClick={[Function]}
        style={
          Object {
            "color": "black",
          }
        }
        testID="Menu.button.1"
        type="button"
      >
        Option 2
      </button>
    </li>
    <li>
      <button
        onClick={[Function]}
        style={
          Object {
            "color": "black",
          }
        }
        testID="Menu.button.2"
        type="button"
      >
        Option 3
      </button>
    </li>
  </ul>
</div>
`;
